# `T20 Cricket: International`

Data source: https://cricsheet.org/downloads/


## Primary Objective:

To build a "Total Runs" prediction model for 1st innings of Men's T20 cricket given any set of values during the match.

# Importing Libraries

In [1]:
import os
from yaml import safe_load

import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

# Collecting the data

In [15]:
filenames = []
for file in os.listdir('/Users/kaushal.totla/Desktop/t20/t20_yamls/t20s'):
    filenames.append(os.path.join('/Users/kaushal.totla/Desktop/t20/t20_yamls/t20s',file))

In [34]:
print("Total files:",len(filenames))

Total files: 2725


In [25]:
print(filenames[:3])

['/Users/kaushal.totla/Desktop/t20/t20_yamls/t20s/1359789.yaml', '/Users/kaushal.totla/Desktop/t20/t20_yamls/t20s/1173055.yaml', '/Users/kaushal.totla/Desktop/t20/t20_yamls/t20s/951375.yaml']


In [22]:
with open(filenames[0],'r') as f:
    df = pd.json_normalize(safe_load(f))

In [23]:
df

innings  meta.data_version  \
0  [{'1st innings': {'team': 'Bermuda', 'deliveri...               0.92   

  meta.created  meta.revision  info.balls_per_over     info.city  \
0   2023-03-01              1                    6  Buenos Aires   

     info.dates info.gender info.match_type  info.match_type_number  ...  \
0  [2023-02-28]        male             T20                    2009  ...   

  info.registry.people.T Manders  info.registry.people.TS Fray  \
0                       6d2b9724                      c7537f8a   

   info.registry.people.VC Ahir info.registry.people.Vijaya Mallela  \
0                      0bb68705                            85b6244a   

  info.registry.people.Z Burgess         info.teams info.toss.winner  \
0                       d0042e16  [Bermuda, Panama]          Bermuda   

  info.toss.decision                info.umpires  \
0                bat  [N Duguid, Vijaya Mallela]   

                                    info.venue  
0  Belgrano Athletic Club Ground, Buenos Aires  

[1 rows x 48 columns]

In [37]:
#Reading all datasets
op = pd.DataFrame()
for file in filenames:
    with open(file,'r') as f:
        df = pd.json_normalize(safe_load(f))
        op = op.append(df)

FileNotFoundError: [Errno 2] No such file or directory: '/Users/kaushal.totla/Desktop/t20/t20_yamls/t20s/README.txt'

Above error is due to the `README.txt` file still being in the folder, we can ignore the error and check if our dataset is prepared.

In [39]:
op.head()

innings  meta.data_version  \
0  [{'1st innings': {'team': 'Bermuda', 'deliveri...               0.92   
0  [{'1st innings': {'team': 'West Indies', 'deli...               0.91   
0  [{'1st innings': {'team': 'India', 'deliveries...               0.91   
0  [{'1st innings': {'team': 'Rwanda', 'deliverie...               0.92   
0  [{'1st innings': {'team': 'Qatar', 'deliveries...               0.92   

  meta.created  meta.revision  info.balls_per_over     info.city  \
0   2023-03-01              1                  6.0  Buenos Aires   
0   2020-03-01              1                  6.0      Canberra   
0   2016-03-17              2                  6.0     Bangalore   
0   2021-08-21              2                  6.0   Kigali City   
0   2022-12-16              1                  6.0       Nairobi   

     info.dates info.gender info.match_type  info.match_type_number  ...  \
0  [2023-02-28]        male             T20                  2009.0  ...   
0  [2020-02-26]      female             T20                   853.0  ...   
0  [2016-03-15]      female             T20                     NaN  ...   
0  [2021-08-21]        male             T20                  1234.0  ...   
0  [2022-12-15]      female             T20                  1322.0  ...   

  info.registry.people.Arshad Mazarzei  info.registry.people.DK Dahani  \
0                                  NaN                             NaN   
0                                  NaN                             NaN   
0                                  NaN                             NaN   
0                                  NaN                             NaN   
0                                  NaN                             NaN   

   info.registry.people.Emran Shahbakhsh info.registry.people.Masood Jayezeh  \
0                                    NaN                                 NaN   
0                                    NaN                                 NaN   
0                                    NaN                                 NaN   
0                                    NaN                                 NaN   
0                                    NaN                                 NaN   

  info.registry.people.Mehran Siasar info.registry.people.Nader Zahadiafzal  \
0                                NaN                                    NaN   
0                                NaN                                    NaN   
0                                NaN                                    NaN   
0                                NaN                                    NaN   
0                                NaN                                    NaN   

  info.registry.people.Naiem Bameri info.registry.people.Navid Abdollahpour  \
0                               NaN                                     NaN   
0                               NaN                                     NaN   
0                               NaN                                     NaN   
0                               NaN                                     NaN   
0                               NaN                                     NaN   

  info.registry.people.Navid Balouch  \
0                                NaN   
0                                NaN   
0                                NaN   
0                                NaN   
0                                NaN   

  info.registry.people.Yousef Shadzehisarjou  
0                                        NaN  
0                                        NaN  
0                                        NaN  
0                                        NaN  
0                                        NaN  

[5 rows x 5065 columns]

Our dataset is ready, let's check the size

In [43]:
#Creating backup
backup = op.copy()

In [105]:
op.shape

(2049, 5065)

- 2049 matches 
- 5065 descriptive columns

# Cleaning the data

In [106]:
op.columns

Index(['innings', 'meta.data_version', 'meta.created', 'meta.revision',
       'info.balls_per_over', 'info.city', 'info.dates', 'info.gender',
       'info.match_type', 'info.match_type_number',
       ...
       'info.registry.people.Arshad Mazarzei',
       'info.registry.people.DK Dahani',
       'info.registry.people.Emran Shahbakhsh',
       'info.registry.people.Masood Jayezeh',
       'info.registry.people.Mehran Siasar',
       'info.registry.people.Nader Zahadiafzal',
       'info.registry.people.Naiem Bameri',
       'info.registry.people.Navid Abdollahpour',
       'info.registry.people.Navid Balouch',
       'info.registry.people.Yousef Shadzehisarjou'],
      dtype='object', length=5065)

In [107]:
#Creating uique index
op.reset_index(drop=True, inplace=True)

In [108]:
#Filtering only for male
op = op[op['info.gender']=='male']

In [109]:
op.drop(op.columns[op.columns.str.startswith('info.registry.people')], axis=1, inplace=True)

In [110]:
pd.Series(op.columns).apply(lambda x: '.'.join(x.split('.')[:2])).unique()

array(['innings', 'meta.data_version', 'meta.created', 'meta.revision',
       'info.balls_per_over', 'info.city', 'info.dates', 'info.gender',
       'info.match_type', 'info.match_type_number', 'info.outcome',
       'info.overs', 'info.player_of_match', 'info.players', 'info.teams',
       'info.toss', 'info.umpires', 'info.venue', 'info.neutral_venue',
       'info.bowl_out'], dtype=object)

In [111]:
op.head(2)

innings  meta.data_version  \
0  [{'1st innings': {'team': 'Bermuda', 'deliveri...               0.92   
3  [{'1st innings': {'team': 'Rwanda', 'deliverie...               0.92   

  meta.created  meta.revision  info.balls_per_over     info.city  \
0   2023-03-01              1                  6.0  Buenos Aires   
3   2021-08-21              2                  6.0   Kigali City   

     info.dates info.gender info.match_type  info.match_type_number  ...  \
0  [2023-02-28]        male             T20                  2009.0  ...   
3  [2021-08-21]        male             T20                  1234.0  ...   

  info.players.Seychelles  info.players.Eswatini  info.players.China  \
0                     NaN                    NaN                 NaN   
3                     NaN                    NaN                 NaN   

  info.players.ICC World XI info.players.Philippines info.toss.uncontested  \
0                       NaN                      NaN                   NaN   
3                       NaN                      NaN                   NaN   

  info.players.South Korea info.players.Gambia info.players.Israel  \
0                      NaN                 NaN                 NaN   
3                      NaN                 NaN                 NaN   

  info.players.Iran  
0               NaN  
3               NaN  

[2 rows x 126 columns]

In [112]:
op['info.overs'].value_counts()

20    1397
Name: info.overs, dtype: int64

In [113]:
op['info.match_type'].value_counts()

T20    1397
Name: info.match_type, dtype: int64

In [114]:
op.drop(['info.gender','info.overs','info.match_type','info.balls_per_over',
         'info.toss.winner','info.toss.decision','meta.data_version','meta.created',
         'meta.revision','info.outcome.bowl_out','info.bowl_out',
         'info.outcome.eliminator','info.outcome.result','info.outcome.method','info.neutral_venue',
         'info.match_type_number','info.outcome.by.runs','info.outcome.by.wickets','info.outcome.winner',
         'info.player_of_match','info.umpires','info.toss.uncontested'], axis=1, inplace=True)

In [115]:
op.drop(op.columns[op.columns.str.startswith('info.players')], axis=1, inplace=True)
op.shape

(1397, 5)

In [116]:
op.head()

innings     info.city  \
0  [{'1st innings': {'team': 'Bermuda', 'deliveri...  Buenos Aires   
3  [{'1st innings': {'team': 'Rwanda', 'deliverie...   Kigali City   
5  [{'1st innings': {'team': 'Netherlands', 'deli...      Kirtipur   
6  [{'1st innings': {'team': 'West Indies', 'deli...    Lauderhill   
7  [{'1st innings': {'team': 'Malta', 'deliveries...         Marsa   

     info.dates               info.teams  \
0  [2023-02-28]        [Bermuda, Panama]   
3  [2021-08-21]          [Rwanda, Ghana]   
5  [2021-04-21]  [Netherlands, Malaysia]   
6  [2019-08-03]     [India, West Indies]   
7  [2021-10-24]     [Malta, Switzerland]   

                                          info.venue  
0        Belgrano Athletic Club Ground, Buenos Aires  
3      Gahanga International Cricket Stadium. Rwanda  
5  Tribhuvan University International Cricket Gro...  
6  Central Broward Regional Park Stadium Turf Ground  
7                                  Marsa Sports Club

In [117]:
op = op.reset_index(drop=True)

In [118]:
op.head(2)

innings     info.city  \
0  [{'1st innings': {'team': 'Bermuda', 'deliveri...  Buenos Aires   
1  [{'1st innings': {'team': 'Rwanda', 'deliverie...   Kigali City   

     info.dates         info.teams  \
0  [2023-02-28]  [Bermuda, Panama]   
1  [2021-08-21]    [Rwanda, Ghana]   

                                      info.venue  
0    Belgrano Athletic Club Ground, Buenos Aires  
1  Gahanga International Cricket Stadium. Rwanda

In [137]:
op['innings'][0][0]['1st innings'].keys()

dict_keys(['team', 'deliveries'])

In [139]:
op['innings'][0][0]['1st innings']['deliveries']

[{0.1: {'bowler': 'AM Patel',
   'runs': {'extras': 1, 'total': 1, 'batsman': 0},
   'extras': {'wides': 1},
   'non_striker': 'T Manders',
   'batsman': 'KS Leverock'}},
 {0.2: {'bowler': 'AM Patel',
   'runs': {'extras': 0, 'total': 0, 'batsman': 0},
   'non_striker': 'T Manders',
   'batsman': 'KS Leverock'}},
 {0.3: {'bowler': 'AM Patel',
   'runs': {'extras': 0, 'total': 0, 'batsman': 0},
   'non_striker': 'T Manders',
   'batsman': 'KS Leverock'}},
 {0.4: {'bowler': 'AM Patel',
   'runs': {'extras': 1, 'total': 1, 'batsman': 0},
   'extras': {'wides': 1},
   'non_striker': 'T Manders',
   'batsman': 'KS Leverock'}},
 {0.5: {'bowler': 'AM Patel',
   'runs': {'extras': 0, 'total': 0, 'batsman': 0},
   'non_striker': 'T Manders',
   'batsman': 'KS Leverock'}},
 {0.6: {'bowler': 'AM Patel',
   'runs': {'extras': 0, 'total': 4, 'batsman': 4},
   'non_striker': 'T Manders',
   'batsman': 'KS Leverock'}},
 {0.7: {'bowler': 'AM Patel',
   'runs': {'extras': 1, 'total': 1, 'batsman': 0},


In [140]:
op['innings'][0][0]['1st innings']['team']

'Bermuda'

We get most of the data needed from this stage of the JSON.

In [150]:
match = op['innings'][0][0]['1st innings']
match_cnt = 1

In [145]:
list(op['innings'][0][0]['1st innings']['deliveries'][0].keys())[0]

0.1

In [148]:
op['innings'][0][0]['1st innings']

{'team': 'Bermuda',
 'deliveries': [{0.1: {'bowler': 'AM Patel',
    'runs': {'extras': 1, 'total': 1, 'batsman': 0},
    'extras': {'wides': 1},
    'non_striker': 'T Manders',
    'batsman': 'KS Leverock'}},
  {0.2: {'bowler': 'AM Patel',
    'runs': {'extras': 0, 'total': 0, 'batsman': 0},
    'non_striker': 'T Manders',
    'batsman': 'KS Leverock'}},
  {0.3: {'bowler': 'AM Patel',
    'runs': {'extras': 0, 'total': 0, 'batsman': 0},
    'non_striker': 'T Manders',
    'batsman': 'KS Leverock'}},
  {0.4: {'bowler': 'AM Patel',
    'runs': {'extras': 1, 'total': 1, 'batsman': 0},
    'extras': {'wides': 1},
    'non_striker': 'T Manders',
    'batsman': 'KS Leverock'}},
  {0.5: {'bowler': 'AM Patel',
    'runs': {'extras': 0, 'total': 0, 'batsman': 0},
    'non_striker': 'T Manders',
    'batsman': 'KS Leverock'}},
  {0.6: {'bowler': 'AM Patel',
    'runs': {'extras': 0, 'total': 4, 'batsman': 4},
    'non_striker': 'T Manders',
    'batsman': 'KS Leverock'}},
  {0.7: {'bowler': 'AM

In [155]:
match['deliveries']

[{0.1: {'bowler': 'AM Patel',
   'runs': {'extras': 1, 'total': 1, 'batsman': 0},
   'extras': {'wides': 1},
   'non_striker': 'T Manders',
   'batsman': 'KS Leverock'}},
 {0.2: {'bowler': 'AM Patel',
   'runs': {'extras': 0, 'total': 0, 'batsman': 0},
   'non_striker': 'T Manders',
   'batsman': 'KS Leverock'}},
 {0.3: {'bowler': 'AM Patel',
   'runs': {'extras': 0, 'total': 0, 'batsman': 0},
   'non_striker': 'T Manders',
   'batsman': 'KS Leverock'}},
 {0.4: {'bowler': 'AM Patel',
   'runs': {'extras': 1, 'total': 1, 'batsman': 0},
   'extras': {'wides': 1},
   'non_striker': 'T Manders',
   'batsman': 'KS Leverock'}},
 {0.5: {'bowler': 'AM Patel',
   'runs': {'extras': 0, 'total': 0, 'batsman': 0},
   'non_striker': 'T Manders',
   'batsman': 'KS Leverock'}},
 {0.6: {'bowler': 'AM Patel',
   'runs': {'extras': 0, 'total': 4, 'batsman': 4},
   'non_striker': 'T Manders',
   'batsman': 'KS Leverock'}},
 {0.7: {'bowler': 'AM Patel',
   'runs': {'extras': 1, 'total': 1, 'batsman': 0},


In [161]:
op.loc[0]

innings       [{'1st innings': {'team': 'Bermuda', 'deliveri...
info.city                                          Buenos Aires
info.dates                                         [2023-02-28]
info.teams                                    [Bermuda, Panama]
info.venue          Belgrano Athletic Club Ground, Buenos Aires
Name: 0, dtype: object

In [163]:
match['deliveries'][0]

{0.1: {'bowler': 'AM Patel',
  'runs': {'extras': 1, 'total': 1, 'batsman': 0},
  'extras': {'wides': 1},
  'non_striker': 'T Manders',
  'batsman': 'KS Leverock'}}

In [315]:
count = 0
t20_data = pd.DataFrame()
for index, data in op.iterrows():

    match_idx = []
    ball_cnt = []
    batsman = []
    bowler = []
    runs = []
    player_out = []
    teams = []
    batting_team = []
    city = []
    venue = []

    for d in data['innings'][0]['1st innings']['deliveries']:
        for key in d.keys():
            
            match_idx.append(count)
            ball_cnt.append(key)
            
            batsman.append(d[key]['batsman'])
            bowler.append(d[key]['bowler'])
            
            runs.append(d[key]['runs']['total'])
            
            try:
                player_out.append(d[key]['wicket']['player_out'])
            except:
                player_out.append('0')
            
            teams.append(data['info.teams'])
            batting_team.append(data['innings'][0]['1st innings']['team'])

            city.append(data['info.city'])
            venue.append(data['info.venue'])
            
    temp = pd.DataFrame({
        'match_idx':match_idx,'ball_cnt':ball_cnt,'batsman':batsman,'bowler':bowler,'runs':runs,
        'player_out':player_out,'teams':teams,'batting_team':batting_team,'city':city,'venue':venue
        })
    
    t20_data = t20_data.append(temp)
    count+=1

In [316]:
t20_data.head(2)

match_idx  ball_cnt      batsman    bowler  runs player_out  \
0          0       0.1  KS Leverock  AM Patel     1          0   
1          0       0.2  KS Leverock  AM Patel     0          0   

               teams batting_team          city  \
0  [Bermuda, Panama]      Bermuda  Buenos Aires   
1  [Bermuda, Panama]      Bermuda  Buenos Aires   

                                         venue  
0  Belgrano Athletic Club Ground, Buenos Aires  
1  Belgrano Athletic Club Ground, Buenos Aires

In [317]:
t20_data.tail(2)

match_idx  ball_cnt        batsman      bowler  runs     player_out  \
120       1396      19.5  Mohammad Amir  TG Southee     2              0   
121       1396      19.6  Mohammad Amir  TG Southee     0  Mohammad Amir   

                       teams batting_team city  \
120  [New Zealand, Pakistan]     Pakistan  NaN   
121  [New Zealand, Pakistan]     Pakistan  NaN   

                                   venue  
120  Dubai International Cricket Stadium  
121  Dubai International Cricket Stadium

In [318]:
def bowling_team_find(x):
    for i in x['teams']:
        if i!=x['batting_team']:
            return i

In [319]:
t20_data['bowling_team'] = t20_data.apply(bowling_team_find, axis=1)

In [320]:
t20_data.drop('teams', axis=1, inplace=True)

In [321]:
t20_data.isnull().sum()*100/t20_data.shape[0]

match_idx       0.000000
ball_cnt        0.000000
batsman         0.000000
bowler          0.000000
runs            0.000000
player_out      0.000000
batting_team    0.000000
city            8.047421
venue           0.000000
bowling_team    0.000000
dtype: float64

- 8% data missing in cities

In [322]:
t20_data[t20_data.city.isnull()]['venue'].value_counts()

Dubai International Cricket Stadium        4783
Harare Sports Club                         2478
Pallekele International Cricket Stadium    1569
Sharjah Cricket Stadium                    1245
Melbourne Cricket Ground                   1079
Sydney Cricket Ground                       623
Sylhet Stadium                              493
Mombasa Sports Club Ground                  371
Adelaide Oval                               371
Moara Vlasiei Cricket Ground                323
Rawalpindi Cricket Stadium                  245
Carrara Oval                                 64
Name: venue, dtype: int64

All the missing values in `city` is present as the ___First word of `venue`___

In [323]:
t20_data['city'] = np.where(t20_data['city'].isnull(), t20_data['venue'].str.split(' ').apply(lambda x: x[0]), t20_data['city'])

In [324]:
t20_data.isnull().sum()

match_idx       0
ball_cnt        0
batsman         0
bowler          0
runs            0
player_out      0
batting_team    0
city            0
venue           0
bowling_team    0
dtype: int64

In [325]:
teams = t20_data['batting_team'].value_counts()[:11].index.to_list()
teams

['Pakistan',
 'India',
 'New Zealand',
 'Sri Lanka',
 'South Africa',
 'West Indies',
 'Australia',
 'England',
 'Afghanistan',
 'Bangladesh',
 'Zimbabwe']

- We will take only major teams and ignore the rest.

In [326]:
t20_data = t20_data[t20_data['batting_team'].isin(teams)]
t20_data = t20_data[t20_data['bowling_team'].isin(teams)]

In [327]:
t20_data.shape

(73585, 10)

In [328]:
t20_data.head()

match_idx  ball_cnt      batsman             bowler  runs   player_out  \
0          3       0.1  JD Campbell  Washington Sundar     0            0   
1          3       0.2  JD Campbell  Washington Sundar     0  JD Campbell   
2          3       0.3     N Pooran  Washington Sundar     0            0   
3          3       0.4     N Pooran  Washington Sundar     1            0   
4          3       0.5      E Lewis  Washington Sundar     0            0   

  batting_team        city                                              venue  \
0  West Indies  Lauderhill  Central Broward Regional Park Stadium Turf Ground   
1  West Indies  Lauderhill  Central Broward Regional Park Stadium Turf Ground   
2  West Indies  Lauderhill  Central Broward Regional Park Stadium Turf Ground   
3  West Indies  Lauderhill  Central Broward Regional Park Stadium Turf Ground   
4  West Indies  Lauderhill  Central Broward Regional Park Stadium Turf Ground   

  bowling_team  
0        India  
1        India  
2        India  
3        India  
4        India

- We will treat this as final dataset.

In [329]:
t20_data.to_csv('cleaned_t20.csv')

# Data Exploration & Feature Engineering

In [507]:
df = t20_data.copy()

In [508]:
df.head()

match_idx  ball_cnt      batsman             bowler  runs   player_out  \
0          3       0.1  JD Campbell  Washington Sundar     0            0   
1          3       0.2  JD Campbell  Washington Sundar     0  JD Campbell   
2          3       0.3     N Pooran  Washington Sundar     0            0   
3          3       0.4     N Pooran  Washington Sundar     1            0   
4          3       0.5      E Lewis  Washington Sundar     0            0   

  batting_team        city                                              venue  \
0  West Indies  Lauderhill  Central Broward Regional Park Stadium Turf Ground   
1  West Indies  Lauderhill  Central Broward Regional Park Stadium Turf Ground   
2  West Indies  Lauderhill  Central Broward Regional Park Stadium Turf Ground   
3  West Indies  Lauderhill  Central Broward Regional Park Stadium Turf Ground   
4  West Indies  Lauderhill  Central Broward Regional Park Stadium Turf Ground   

  bowling_team  
0        India  
1        India  
2        India  
3        India  
4        India

In [509]:
df.shape

(73585, 10)

In [510]:
#No. of unique values
df.nunique()

match_idx       606
ball_cnt        181
batsman         681
bowler          539
runs              8
player_out      606
batting_team     11
city             95
venue           141
bowling_team     11
dtype: int64

- 606 matches across 95 cities and 141 venues.
- Total 11 teams

There are many cities, we will take top 40 cities and ignore the rest.

In [511]:
df['city'].value_counts()[:40]

Dubai              4024
Harare             3589
Colombo            3464
Mirpur             2563
Johannesburg       2509
Abu Dhabi          2112
Auckland           2033
Sydney             1950
Lahore             1858
Dhaka              1826
Melbourne          1699
Sharjah            1616
Cape Town          1615
Pallekele          1569
St Lucia           1358
Hamilton           1340
Durban             1273
Nottingham         1241
Centurion          1129
Southampton        1116
London             1115
Karachi            1112
Barbados           1051
Lauderhill          995
Christchurch        988
Wellington          977
Manchester          918
Perth               867
Ahmedabad           861
Kolkata             860
Brisbane            845
Gros Islet          773
Lucknow             754
Mount Maunganui     754
Chandigarh          751
Mumbai              747
Adelaide            742
Delhi               742
Cardiff             733
Bridgetown          685
Name: city, dtype: int64

In [512]:
df = df[df['city'].isin(df['city'].value_counts()[:40].index)]

In [513]:
#player out column also has many unique values which won't be very useful for the model. we will perform flagging on this.
df['player_out'] = np.where(df['player_out']=='0',0,1)

In [514]:
df['player_out'].value_counts()

0    54132
1     3022
Name: player_out, dtype: int64

In [515]:
cumulatives = df.groupby('match_idx').cumsum()
cumulatives

ball_cnt  runs  player_out
0         0.1     0           0
1         0.3     0           1
2         0.6     0           1
3         1.0     1           1
4         1.5     1           1
..        ...   ...         ...
117    1123.6   158           6
118    1142.9   158           7
119    1162.3   159           7
120    1181.8   161           7
121    1201.4   161           8

[57154 rows x 3 columns]

We will change `runs` and `player_out` as abovev for continual information.

In [516]:
df['score'] = cumulatives['runs']
df['player_out'] = cumulatives['player_out']

In [517]:
#balls and overs
df['ball_cnt']

0       0.1
1       0.2
2       0.3
3       0.4
4       0.5
       ... 
117    19.2
118    19.3
119    19.4
120    19.5
121    19.6
Name: ball_cnt, Length: 57154, dtype: float64

1st part is over, second part is balls in that over

In [518]:
df['overs'] = df['ball_cnt'].astype('str').str.split('.').apply(lambda x:x[0]).astype(int)
df['balls'] = df['ball_cnt'].astype('str').str.split('.').apply(lambda x:x[1]).astype(int)

In [519]:
df[['overs','balls']].describe()

overs         balls
count  57154.000000  57154.000000
mean       9.400602      3.622599
std        5.760721      1.808484
min        0.000000      1.000000
25%        4.000000      2.000000
50%        9.000000      4.000000
75%       14.000000      5.000000
max       19.000000     12.000000

In [520]:
df['balls_left'] = 120 - (df['overs']*6 + df['balls'])
df['balls_left'].describe()

count    57154.000000
mean        59.973790
std         34.611087
min         -3.000000
25%         30.000000
50%         60.000000
75%         90.000000
max        119.000000
Name: balls_left, dtype: float64

In [521]:
df['balls_left'] = df['balls_left'].apply(lambda x: 0 if x<0 else x)

In [522]:
df['wickets_left'] = 10 - df['player_out']
df['wickets_left'].describe()

count    57154.000000
mean         7.383630
std          2.141368
min          0.000000
25%          6.000000
50%          8.000000
75%          9.000000
max         10.000000
Name: wickets_left, dtype: float64

In [523]:
df['score'].describe()

count    57154.000000
mean        75.997813
std         49.885309
min          0.000000
25%         35.000000
50%         71.000000
75%        112.000000
max        258.000000
Name: score, dtype: float64

In [524]:
df['crr'] = df['score']/(df['overs']*6 + df['balls'])
df['crr'].describe()

count    57154.000000
mean         1.216507
std          0.389875
min          0.000000
25%          1.000000
50%          1.223177
75%          1.441176
max         14.000000
Name: crr, dtype: float64

In [525]:
df['powerplay'] = np.where(df['overs']<=6,1,0)
df['death_overs'] = np.where(df['overs']>=15,1,0)
df[['powerplay','death_overs']].describe()

powerplay   death_overs
count  57154.000000  57154.000000
mean       0.356388      0.244287
std        0.478936      0.429668
min        0.000000      0.000000
25%        0.000000      0.000000
50%        0.000000      0.000000
75%        1.000000      0.000000
max        1.000000      1.000000

In [526]:
df['top_order_batsmen'] = np.where(df['player_out']<3,1,0)
df['middle_order_batsmen'] = np.where((df['player_out']>=3) & (df['player_out']<=4),1,0)
df['lower_order_batsmen'] = np.where((df['player_out']>=5) & (df['player_out']<=7),1,0)
df['tail_order_batsmen'] = np.where(df['player_out']>=8,1,0)

In [527]:
df['pressure_play'] = np.where((df['crr']<7)&(df['overs']>=2),1,0)
#when crr is <7 and beyond 2 overs

In [532]:
match_groups = df.groupby('match_idx')
mids = df['match_idx'].unique()
l5overs = []
for idx in mids:
    l5overs.extend((match_groups.get_group(idx).rolling(min_periods=1, window=30).sum()['runs']).to_list())

In [534]:
df['last_5_overs_score'] = l5overs

In [535]:
df['agressive'] = np.where(df['last_5_overs_score']>=45,1,0)
df['agressive'].value_counts()

0    44158
1    12996
Name: agressive, dtype: int64

In [537]:
pd.set_option('display.max_columns',50)

In [539]:
df.head(2)

match_idx  ball_cnt      batsman             bowler  runs  player_out  \
0          3       0.1  JD Campbell  Washington Sundar     0           0   
1          3       0.2  JD Campbell  Washington Sundar     0           1   

  batting_team        city                                              venue  \
0  West Indies  Lauderhill  Central Broward Regional Park Stadium Turf Ground   
1  West Indies  Lauderhill  Central Broward Regional Park Stadium Turf Ground   

  bowling_team  score  overs  balls  balls_left  wickets_left  crr  powerplay  \
0        India      0      0      1         119            10  0.0          1   
1        India      0      0      2         118             9  0.0          1   

   death_overs  top_order_batsmen  middle_order_batsmen  lower_order_batsmen  \
0            0                  1                     0                    0   
1            0                  1                     0                    0   

   tail_order_batsmen  pressure_play  last_5_overs_score  agressive  
0                   0              0                 0.0          0  
1                   0              0                 0.0          0

### Building Target variable

Our target variable would be Total Score per match.

In [541]:
totals = df.groupby('match_idx')['runs'].sum()

In [544]:
df = pd.merge(df, totals, how='left', on='match_idx').rename(columns={'runs_x':'runs', 'runs_y':'total_runs'})

In [545]:
df.columns

Index(['match_idx', 'ball_cnt', 'batsman', 'bowler', 'runs', 'player_out',
       'batting_team', 'city', 'venue', 'bowling_team', 'score', 'overs',
       'balls', 'balls_left', 'wickets_left', 'crr', 'powerplay',
       'death_overs', 'top_order_batsmen', 'middle_order_batsmen',
       'lower_order_batsmen', 'tail_order_batsmen', 'pressure_play',
       'last_5_overs_score', 'agressive', 'total_runs'],
      dtype='object')

In [548]:
df.drop(['match_idx','ball_cnt','batsman','bowler','runs','player_out','balls','overs'], axis=1, inplace=True)

In [713]:
df.to_csv('t2_cleaned.csv')

In [151]:
df = pd.read_csv('t2_cleaned.csv').drop('Unnamed: 0', axis=1)

In [152]:
df.columns

Index(['batting_team', 'city', 'venue', 'bowling_team', 'score', 'balls_left',
       'wickets_left', 'crr', 'powerplay', 'death_overs', 'top_order_batsmen',
       'middle_order_batsmen', 'lower_order_batsmen', 'tail_order_batsmen',
       'pressure_play', 'last_5_overs_score', 'agressive', 'total_runs'],
      dtype='object')

# Data Preparation for Modelling

In [153]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

In [154]:
x = df.drop(['total_runs','city'], axis=1)
y = df['total_runs']

In [155]:
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.2, random_state=10)

In [156]:
x_train.shape, x_test.shape

((45723, 16), (11431, 16))

In [157]:
x_train['batting_team'].value_counts()

Pakistan        6770
India           6004
New Zealand     5537
Sri Lanka       4549
Australia       4507
South Africa    4165
West Indies     3995
England         3613
Bangladesh      2661
Afghanistan     1985
Zimbabwe        1937
Name: batting_team, dtype: int64

In [158]:
x_test['batting_team'].value_counts()

Pakistan        1723
India           1430
New Zealand     1347
Sri Lanka       1144
Australia       1123
West Indies     1029
South Africa     991
England          857
Bangladesh       762
Zimbabwe         516
Afghanistan      509
Name: batting_team, dtype: int64

In [159]:
cat_cols = ['batting_team', 'bowling_team', 'venue'] 
num_cols = ['score', 'balls_left','wickets_left', 'crr', 'last_5_overs_score']
binary_cols = list(set(df.columns).difference(cat_cols).difference(num_cols).difference({'total_runs', 'city'}))

In [160]:
transformations = ColumnTransformer([
    ('cols',OneHotEncoder(drop='first',sparse=False),cat_cols),
    ('nums',StandardScaler(),num_cols)
], remainder='passthrough')

In [161]:
cols = pd.get_dummies(x_train,drop_first=True).columns

In [162]:
ohe = OneHotEncoder(sparse=False, drop='first')
x_train_cats = ohe.fit_transform(x_train[cat_cols])
x_test_cats = ohe.transform(x_test[cat_cols])

In [163]:
scale = StandardScaler()
x_train_nums = scale.fit_transform(x_train[num_cols])
x_test_nums = scale.transform(x_test[num_cols])

In [164]:
x_train = pd.DataFrame(np.hstack([x_train_cats,x_train_nums,x_train[binary_cols]]), columns=cols)
x_test = pd.DataFrame(np.hstack([x_test_cats,x_test_nums,x_test[binary_cols]]), columns=cols)

In [165]:
x_train.shape

(45723, 104)

# Modelling

- Due to huge size of the dataset, we will avoid models which use __distances__.
- Due to low latency requirement, we will also avoid SVMs

We will consider using below models:
- LinearRegression #As base model (Also, Lasso, Ridge) 
- RandomForestRegressor
- GBRegressor
- XGBRegressor


Performance metrics: `R2` and `RMSE`

In [179]:
from sklearn.linear_model import Lasso,Ridge,LinearRegression
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from xgboost import XGBRegressor

from sklearn.metrics import r2_score, mean_squared_error as mse

#### Linear Regression

In [167]:
lr = LinearRegression()
lr.fit(x_train, y_train)
y_pred_train = lr.predict(x_train)
y_pred_test = lr.predict(x_test)

In [168]:
print("Train:")
print(r2_score(y_train, y_pred_train))
print(mse(y_train, y_pred_train)**(0.5))

print("Test:")
print(r2_score(y_test, y_pred_test))
print(mse(y_test, y_pred_test)**(0.5))

Train:
0.6380202689845894
19.813678282088546
Test:
0.6388238455357558
19.941524480820913


- Very low scores with Lienar Regression. Adding Regulariations will make it worse.

In [169]:
pd.DataFrame({'cols':lr.feature_names_in_, 'fi':lr.coef_}).sort_values(by='fi', ascending=False).values

array([['bowling_team_West Indies', 14753865819324.963],
       ['bowling_team_England', 14753865819315.13],
       ['bowling_team_India', 14753865819310.506],
       ['bowling_team_Zimbabwe', 14753865819310.406],
       ['venue_Western Australia Cricket Association Ground',
        38.986594704309766],
       ['venue_Melbourne Cricket Ground', 34.81682542880526],
       ['venue_The Wanderers Stadium', 29.387366075909448],
       ['venue_Pallekele International Cricket Stadium',
        27.065202371059076],
       ['venue_Westpac Stadium', 26.59208145125188],
       ['venue_Perth Stadium', 24.74730692654883],
       ['venue_Sophia Gardens, Cardiff', 17.465750344575394],
       ['batting_team_Zimbabwe', 16.15667626944049],
       ['venue_R Premadasa Stadium', 13.620143235040121],
       ['venue_Zayed Cricket Stadium, Abu Dhabi', 13.33855790663393],
       ['venue_Sydney Cricket Ground', 12.20969208069782],
       ['venue_Feroz Shah Kotla', 11.644143408138238],
       ['wickets_left', 10

#### RandomForest

In [170]:
rf = RandomForestRegressor()
rf.fit(x_train, y_train)
y_pred_train = rf.predict(x_train)
y_pred_test = rf.predict(x_test)

In [171]:
print("Train:")
print(r2_score(y_train, y_pred_train))
print(mse(y_train, y_pred_train)**(0.5))

print("Test:")
print(r2_score(y_test, y_pred_test))
print(mse(y_test, y_pred_test)**(0.5))

Train:
0.9926420443762971
2.8248901391476786
Test:
0.9586957722874723
6.743666340262158


- very significant performance increase with RF. We will consider it for hypertuning.

In [172]:
pd.DataFrame({'cols':rf.feature_names_in_, 'fi':rf.feature_importances_}).sort_values(by='fi', ascending=False).values

array([['bowling_team_Australia', 0.4570635777034105],
       ['venue_Zayed Cricket Stadium, Abu Dhabi', 0.08427256837484906],
       ['venue_Western Australia Cricket Association Ground',
        0.04326262001378388],
       ['bowling_team_Bangladesh', 0.016774667044443067],
       ['venue_Westpac Stadium', 0.014934098834583014],
       ['score', 0.012442213536319588],
       ['crr', 0.01222109302091514],
       ['venue_Central Broward Regional Park Stadium Turf Ground, Lauderhill',
        0.01139777195207972],
       ['pressure_play', 0.010845021518973914],
       ['batting_team_Pakistan', 0.009899392930852312],
       ['venue_Harare Sports Club', 0.00924033595856732],
       ['batting_team_New Zealand', 0.009220466737683286],
       ['balls_left', 0.00920450649229287],
       ['agressive', 0.009134017231114474],
       ['middle_order_batsmen', 0.008964799649551297],
       ['top_order_batsmen', 0.008693380490800854],
       ['lower_order_batsmen', 0.008619887263092168],
       ['la

#### GBDT

In [680]:
gb = GradientBoostingRegressor()
gb.fit(x_train, y_train)
y_pred_train = gb.predict(x_train)
y_pred_test = gb.predict(x_test)

In [681]:
print("Train:")
print(r2_score(y_train, y_pred_train))
print(mse(y_train, y_pred_train)**(0.5))

print("Test:")
print(r2_score(y_test, y_pred_test))
print(mse(y_test, y_pred_test)**(0.5))

Train:
0.6833175131855541
18.532542461666914
Test:
0.681831072028187
18.71663208709132


- Horrible performance. Almost equivalent to that of `linear regression`

#### XGBRegressor

In [173]:
xgb = XGBRegressor()
xgb.fit(x_train, y_train)
y_pred_train = xgb.predict(x_train)
y_pred_test = xgb.predict(x_test)

In [174]:
print("Train:")
print(r2_score(y_train, y_pred_train))
print(mse(y_train, y_pred_train)**(0.5))

print("Test:")
print(r2_score(y_test, y_pred_test))
print(mse(y_test, y_pred_test)**(0.5))

Train:
0.9193487844078219
9.352515557281746
Test:
0.9066156988804589
10.13994869554122


- Very fast and reliable.
- Good performance in terms of R2. Slightly off in terms of RMSE.

# Hyperparameter Tuning

In [185]:
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import make_scorer

scores = {'mse': make_scorer(mse),'r2': make_scorer(r2_score)}

#### RF

In [221]:
rf_gs = RandomForestRegressor(n_jobs=-1)

#Random forest
rf_params_grid = {'n_estimators':[70,100,150,200],
              'max_features':['sqrt','log2',None],
             'max_samples':[0.2,0.5,0.7,1]}

In [222]:
rf_gscv = GridSearchCV(rf_gs, rf_params_grid, cv=3, scoring=scores, refit='r2')
rf_gscv.fit(x_train, y_train)

GridSearchCV(cv=3, estimator=RandomForestRegressor(n_jobs=-1),
             param_grid={'max_features': ['sqrt', 'log2', None],
                         'max_samples': [0.2, 0.5, 0.7, 1],
                         'n_estimators': [70, 100, 150, 200]},
             refit='r2',
             scoring={'mse': make_scorer(mean_squared_error),
                      'r2': make_scorer(r2_score)})

In [223]:
pd.DataFrame(rf_gscv.cv_results_).sort_values(by='mean_test_r2', ascending=False)

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
26       1.262482      0.028150         0.085509        0.009098   
27       1.628265      0.014726         0.099190        0.003579   
25       0.885217      0.014902         0.057257        0.000835   
24       0.736479      0.047674         0.046156        0.001622   
42       5.461267      0.215524         0.092678        0.008668   
43       6.957753      0.188505         0.108909        0.006216   
41       3.726858      0.144205         0.063615        0.010249   
40       2.489907      0.106815         0.048093        0.003732   
11       1.777338      0.046224         0.103659        0.000195   
10       1.454341      0.040615         0.101950        0.011297   
9        1.049750      0.018152         0.059922        0.001516   
8        0.862186      0.022697         0.047876        0.003277   
23       1.312886      0.025826         0.096608        0.000794   
21       0.923166      0.014581         0.057311        0.001709   
22       1.115577      0.059765         0.082032        0.002168   
20       0.725299      0.104203         0.047572        0.002712   
38       3.660653      0.072216         0.080071        0.005431   
39       4.897890      0.072483         0.098839        0.005505   
36       1.823830      0.020663         0.042959        0.001034   
37       2.589851      0.063034         0.056391        0.002075   
7        1.653299      0.066655         0.139593        0.022970   
6        1.324467      0.034552         0.103347        0.009966   
4        0.790043      0.062891         0.056054        0.012267   
5        0.910718      0.019914         0.069614        0.007529   
19       1.168208      0.196308         0.105833        0.015776   
17       0.586051      0.035156         0.061048        0.009889   
18       0.809100      0.029036         0.070610        0.000918   
16       0.521713      0.104380         0.039513        0.000516   
3        0.910639      0.047869         0.086259        0.002988   
35       2.232589      0.040018         0.092179        0.009171   
34       1.750280      0.090981         0.069193        0.002842   
2        0.695274      0.005955         0.069228        0.005901   
32       0.836059      0.034088         0.038925        0.001117   
0        0.456861      0.003009         0.041259        0.001436   
1        0.551956      0.006868         0.056609        0.010245   
33       1.106056      0.079351         0.043359        0.003910   
28       0.258880      0.016759         0.023049        0.001320   
15       0.377263      0.051854         0.045185        0.001224   
13       0.382982      0.051041         0.026468        0.000324   
45       0.262375      0.001587         0.027525        0.000422   
31       0.414337      0.079185         0.041851        0.000570   
44       0.332231      0.121644         0.026287        0.003650   
29       0.277750      0.025329         0.026015        0.000261   
30       0.294138      0.008572         0.033647        0.000444   
46       0.302252      0.008024         0.033615        0.000486   
12       0.302143      0.049724         0.022475        0.002019   
14       0.353844      0.031817         0.035478        0.001394   
47       0.334129      0.002509         0.043275        0.002337   

   param_max_features param_max_samples param_n_estimators  \
26               log2               0.7                150   
27               log2               0.7                200   
25               log2               0.7                100   
24               log2               0.7                 70   
42               None               0.7                150   
43               None               0.7                200   
41               None               0.7                100   
40               None               0.7                 70   
11               sqrt               0.7                200   
10               sqrt               0.7            

#### xgb

In [183]:
xgb_cv = XGBRegressor(n_jobs=-1)

xgb_params_grid = {'learning_rate':[0.05, 0.1, 0.3],
                   'max_depth':[2,5,10],
                  'min_child_weight':[1,3,5,7],
                  }

In [186]:
xgb_gscv = GridSearchCV(xgb_cv, xgb_params_grid, cv=3, scoring=scores, refit='r2')
xgb_gscv.fit(x_train, y_train)

GridSearchCV(cv=3,
             estimator=XGBRegressor(base_score=None, booster=None,
                                    callbacks=None, colsample_bylevel=None,
                                    colsample_bynode=None,
                                    colsample_bytree=None,
                                    early_stopping_rounds=None,
                                    enable_categorical=False, eval_metric=None,
                                    feature_types=None, gamma=None, gpu_id=None,
                                    grow_policy=None, importance_type=None,
                                    interaction_constraints=None,
                                    learning_rate=None, m...
                                    max_depth=None, max_leaves=None,
                                    min_child_weight=None, missing=nan,
                                    monotone_constraints=None, n_estimators=100,
                                    n_jobs=-1, num_parallel_tree=None,
                                    predictor=None, random_state=None, ...),
             param_grid={'learning_rate': [0.05, 0.1, 0.3],
                         'max_depth': [2, 5, 10],
                         'min_child_weight': [1, 3, 5, 7]},
             refit='r2',
             scoring={'mse': make_scorer(mean_squared_error),
                      'r2': make_scorer(r2_score)})

In [191]:
pd.DataFrame(xgb_gscv.cv_results_).sort_values(by='mean_test_r2', ascending=False)

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
32       4.584996      0.070493         0.016493        0.001056   
33       4.533553      0.031273         0.016765        0.001352   
34       4.531016      0.035999         0.016382        0.000725   
35       4.553567      0.013254         0.016593        0.000765   
20       4.573304      0.006183         0.016226        0.000524   
21       4.549175      0.022608         0.016873        0.000446   
22       4.725133      0.261055         0.016029        0.000763   
23       4.413591      0.033110         0.015791        0.000367   
31       2.311545      0.088378         0.011084        0.000216   
30       2.278632      0.045536         0.011906        0.000943   
29       2.236237      0.004919         0.011124        0.000198   
28       2.262973      0.046250         0.011110        0.000192   
8        5.713548      0.219419         0.017506        0.000975   
9        5.062315      0.331558         0.016750        0.000288   
10       4.720678      0.085158         0.015990        0.000421   
11       4.627607      0.010443         0.016033        0.000270   
16       2.331678      0.025400         0.012489        0.002308   
17       2.243248      0.020246         0.011540        0.000650   
18       2.288301      0.011841         0.011124        0.000088   
19       2.273674      0.007132         0.011014        0.000173   
5        2.281282      0.069075         0.011201        0.000726   
4        2.393235      0.100609         0.015309        0.005680   
7        2.796941      0.281534         0.012846        0.001049   
6        2.793446      0.128041         0.013515        0.001031   
24       0.987362      0.012232         0.008834        0.000094   
25       1.007425      0.028081         0.008726        0.000112   
26       0.989342      0.017282         0.008796        0.000053   
27       0.979147      0.010773         0.008665        0.000025   
15       1.068038      0.006065         0.011496        0.002729   
12       1.017350      0.009682         0.008837        0.000207   
13       1.013893      0.015678         0.008714        0.000013   
14       1.021882      0.008904         0.009100        0.000221   
1        0.885674      0.026771         0.007978        0.000508   
3        1.067983      0.062925         0.009928        0.001434   
2        0.874808      0.013479         0.008689        0.001001   
0        0.891077      0.006813         0.007934        0.000161   

   param_learning_rate param_max_depth param_min_child_weight  \
32                 0.3              10                      1   
33                 0.3              10                      3   
34                 0.3              10                      5   
35                 0.3              10                      7   
20                 0.1              10                      1   
21                 0.1              10                      3   
22                 0.1              10                      5   
23                 0.1              10                      7   
31                 0.3               5                      7   
30                 0.3               5                      5   
29                 0.3               5                      3   
28                 0.3               5                      1   
8                 0.05              10                      1   
9                 0.05              10                      3   
10                0.05              10                      5   
11                0.05              10                      7   
16                 0.1               5                      1   
17                 0.1               5                      3   
18                 0.1               5                      5   
19                 0.1               5                      7   
5                 0.05               5                      3   
4                 0.05               5                   

In [225]:
np.sqrt(50.890076)

7.133728057614756

In [192]:
xgb_gscv

GridSearchCV(cv=3,
             estimator=XGBRegressor(base_score=None, booster=None,
                                    callbacks=None, colsample_bylevel=None,
                                    colsample_bynode=None,
                                    colsample_bytree=None,
                                    early_stopping_rounds=None,
                                    enable_categorical=False, eval_metric=None,
                                    feature_types=None, gamma=None, gpu_id=None,
                                    grow_policy=None, importance_type=None,
                                    interaction_constraints=None,
                                    learning_rate=None, m...
                                    max_depth=None, max_leaves=None,
                                    min_child_weight=None, missing=nan,
                                    monotone_constraints=None, n_estimators=100,
                                    n_jobs=-1, num_parallel_tree=None,
                                    predictor=None, random_state=None, ...),
             param_grid={'learning_rate': [0.05, 0.1, 0.3],
                         'max_depth': [2, 5, 10],
                         'min_child_weight': [1, 3, 5, 7]},
             refit='r2',
             scoring={'mse': make_scorer(mean_squared_error),
                      'r2': make_scorer(r2_score)})

In [195]:
xgb_gscv.best_estimator_

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=None, gpu_id=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=0.3, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=10, max_leaves=None,
             min_child_weight=1, missing=nan, monotone_constraints=None,
             n_estimators=100, n_jobs=-1, num_parallel_tree=None,
             predictor=None, random_state=None, ...)

# Building Pickle Files for deployment

In [210]:
import pickle

In [201]:
cols = ColumnTransformer([
    ('ohe',OneHotEncoder(sparse=False,drop='first'),cat_cols),
    ('scale',StandardScaler(),num_cols)
], remainder='passthrough')

In [202]:
pipe = Pipeline([
    ('transformations',cols),
    ('model',xgb_gscv.best_estimator_)
])

In [206]:
x = df.drop(['city','total_runs'], axis=1)
y = df['total_runs']

In [207]:
pipe.fit(x, y)

Pipeline(steps=[('transformations',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('ohe',
                                                  OneHotEncoder(drop='first',
                                                                sparse=False),
                                                  ['batting_team',
                                                   'bowling_team', 'venue']),
                                                 ('scale', StandardScaler(),
                                                  ['score', 'balls_left',
                                                   'wickets_left', 'crr',
                                                   'last_5_overs_score'])])),
                ('model',
                 XGBRegressor(base_score=None, booster=None, callbacks=None,
                              colsample...
                              feature_types=None, gamma=None, gpu_id=None,
                              grow_policy=None, importance_type=None,
                              interaction_constraints=None, learning_rate=0.3,
                              max_bin=None, max_cat_threshold=None,
                              max_cat_to_onehot=None, max_delta_step=None,
                              max_depth=10, max_leaves=None, min_child_weight=1,
                              missing=nan, monotone_constraints=None,
                              n_estimators=100, n_jobs=-1,
                              num_parallel_tree=None, predictor=None,
                              random_state=None, ...))])

In [211]:
# Exporting model in a pickle file
pickle.dump(pipe, open('pipe.pkl','wb'))

In [212]:
import xgboost

In [213]:
from importlib.metadata import version

In [215]:
version('xgboost')

'1.7.6'